### PR TITLE
chore: balancer CoW AMM milestone

### PR DESCRIPTION
**Pull request by wei3erHase**
To be merged after:
- [x] #102 
- [x] #110 

This PR modifies:
- Adds BCoWPool (inherits BPool and adds functionality)
    - Infinite approval on finalize
    - ERC1271 swap approval
    - Emit event on pool finalization
- Adds BCoWFactory
- Changes reentrancy guard mechanism to transient
- Changes `require` string errors for `CustomErrors()`
- Adds `_controller_` modifier

### CODE REVIEW

File: script/Params.s.sol [hunk 1]
```diff
@@ -2,18 +2,35 @@
...
+  /// @notice BFactory deployment parameters for each chain
+  mapping(uint256 _chainId => BFactoryDeploymentParams _params) internal _bFactoryDeploymentParams;
+
+  /// @notice BCoWFactory deployment parameters for each chain
+  mapping(uint256 _chainId => BCoWFactoryDeploymentParams _params) internal _bCoWFactoryDeploymentParams;
 
   constructor() {
     // Mainnet
-    _deploymentParams[1] = DeploymentParams(address(this));
+    _bFactoryDeploymentParams[1] = BFactoryDeploymentParams(address(this));
```
Comment (wei3erHase) [inline script/Params.s.sol:29]: we can store `address(this)` in a `_B_LABS` constant to be set once

File: src/contracts/BCoWPool.sol [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  }
+
+  /// @inheritdoc IBCoWPool
+  function commit(bytes32 orderHash) external {
+    if (msg.sender != address(SOLUTION_SETTLER)) {
+      revert CommitOutsideOfSettlement();
+    }
+    if (_getLock() != _MUTEX_FREE) {
+      revert BCoWPool_CommitmentAlreadySet();
+    }
```
Comment (wei3erHase) [inline src/contracts/BCoWPool.sol:63]: could be replaced by `_viewlock_` 🤔

File: src/contracts/BFactory.sol [hunk 3]
```diff
@@ -30,17 +30,23 @@ contract BFactory is IBFactory {
...
-    require(msg.sender == _blabs, 'ERR_NOT_BLABS');
+    if (msg.sender != _blabs) {
+      revert BFactory_NotBLabs();
+    }
     uint256 collected = pool.balanceOf(address(this));
     bool xfer = pool.transfer(_blabs, collected);
-    require(xfer, 'ERR_ERC20_FAILED');
+    if (!xfer) {
+      revert BFactory_ERC20TransferFailed();
+    }
```
Comment (wei3erHase) [inline src/contracts/BFactory.sol:49]: should we use SafeERC20 here? wdyt?

File: src/contracts/BPool.sol [hunk 4]
```diff
@@ -36,15 +34,29 @@ contract BPool is BToken, BMath, IBPool {
...
+    if (_getLock() != _MUTEX_FREE) {
+      revert BPool_Reentrancy();
+    }
+    _setLock(_MUTEX_TAKEN);
     _;
-    _mutex = false;
+    _setLock(_MUTEX_FREE);
   }
 
   /// @dev Prevents reentrancy in view functions
```
Comment (wei3erHase) [inline src/contracts/BPool.sol:45]: perhaps could change description to "Throws an error if the reentrancy mutex is locked (doesn't modify the mutex)"

File: src/contracts/BCoWPool.sol (hunk 2 above)
Comment (0xteddybear) [reply src/contracts/BCoWPool.sol:63]: we would get another, less descriptive error message. Other than that, I agree

File: src/contracts/BFactory.sol (hunk 3 above)
Comment (0xteddybear) [reply src/contracts/BFactory.sol:49]: I think it's worth testing wether we'd get an abi-decode error when calling a `transfer` method which doesn't have a return value (which could break collections). SafeERC20 would help with that, but idk if we would want to ignore transfer failures altogether instead

File: script/Params.s.sol (hunk 1 above)
Comment (0xteddybear) [reply script/Params.s.sol:29]: Shouldn't `blabs` be a a runtime parameter? in that way the deployment can be run with keys with a lower security requirements than the one that'll have the `bLabs` role, e.g. `bLabs` could be a multisig and the contracts be deployed by a random dev with some ether

File: test/unit/BCoWFactory.t.sol [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    vm.mockCall(solutionSettler, abi.encodePacked(ISettlement.domainSeparator.selector), abi.encode(bytes32(0)));
+    vm.mockCall(
+      solutionSettler, abi.encodePacked(ISettlement.vaultRelayer.selector), abi.encode(makeAddr('vault relayer'))
+    );
+    vm.prank(owner);
+    return new MockBCoWFactory(solutionSettler, appData);
+  }
+
+  function _bPoolBytecode() internal virtual override returns (bytes memory _bytecode) {
+    vm.skip(true);
```
Comment (0xteddybear) [inline test/unit/BCoWFactory.t.sol:26]: I'm not getting any test skips with the `Deploy_New_BPool`

    > FOUNDRY_FUZZ_RUNS=10 nice forge test --mt Deploy_NewBPool |grep SKIP
    
we are not extending the `BFactory_Internal_NewBPool` contract

also, we could save the fuzzed constructor inputs into storage, deploy the contract here reading them, and then return `address(pool).code`

File: README.md [hunk 6]
```diff
@@ -1,83 +1,57 @@
...
+- Implements a `commit` method to avoid multiple swaps from conflicting with each other.
+  - This is stored in the same transient storage slot as reentrancy locks in order to prevent calls to swap/join functions within a settlement execution or vice versa.
+  - It's an error to override a commitment since that could be used to clear reentrancy locks. Commitments can only be cleared by ending a transaction.
+- Validates the `GPv2Order` requirements before allowing the swap
+
+## Features on BCoWFactory
+- Added a `logBCoWPool` to log the finalization of BCoWPool contracts, to be called by a child pool
+
+## Creating a Pool
+- Create a new pool by calling `IBFactory.newBPool()`
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply README.md:51]: i would def wanna see that 👀

File: src/contracts/BCoWConst.sol [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+   * @return _emptyCommitment The commitment value representing no commitment.
+   */
+  bytes32 public constant EMPTY_COMMITMENT = bytes32(0);
+
+  /**
+   * @notice The largest possible duration of any AMM order, starting from the
+   * current block timestamp.
+   * @return _maxOrderDuration The maximum order duration.
+   */
+  uint32 public constant MAX_ORDER_DURATION = 5 * 60;
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply src/contracts/BCoWConst.sol:20]: isn't it deprecated?

File: src/contracts/BCoWPool.sol [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    _setLock(orderHash);
+  }
+
+  /**
+   * @inheritdoc IERC1271
+   * @dev this function reverts if the order hash does not match the current commitment
+   */
+  function isValidSignature(bytes32 _hash, bytes memory signature) external view returns (bytes4) {
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply src/contracts/BCoWPool.sol:71]: the way we're using ERC1271 is that CoW (Coincidence of Wants) matches 2 orders that sell and buy the same thing, one is from an EOA, that uses the typical signature mechanism, and the counter-order is "signed" by BCoWPool contract.

in that way, i think i agree that the names are confusing, perhaps the 1st argument should be the signature, while the second be the payload (or thing to sign)

the expectation of `isValidSignature` is that if it is indeed valid, the method should return a `MAGIC_NUMBER`, then the contract that has full ERC20 allowance can transfer the tokens in the pool's name

[the remark this answers is not in the document]
Comment (wei3erHase) [reply src/contracts/BCoWPool.sol:71]: the "cryptographic signature" in our case would be a valid `order`, if the order is valid, the pool will sign anything

File: src/interfaces/IBCoWPool.sol [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+   */
+  // solhint-disable-next-line style-guide-casing
+  function APP_DATA() external view returns (bytes32 _appData);
+
+  /**
+   * @notice This function returns the commitment hash that has been set by the
+   * `commit` function. If no commitment has been set, then the value will be
+   * `EMPTY_COMMITMENT`.
+   * @return _commitment The commitment hash.
+   * @dev since commitments share a transient storage slot with reentrancy
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply src/interfaces/IBCoWPool.sol:117]: *Since

[the remark this answers is not in the document]
Comment (wei3erHase) [reply src/interfaces/IBCoWPool.sol:117]: and `EMPTY_COMMITMENT` shouldn't exist anymore 🙈

File: src/contracts/BCoWConst.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (0xteddybear) [reply src/contracts/BCoWConst.sol:20]: only `years` was removed

File: src/interfaces/IBCoWPool.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (wei3erHase) [reply src/interfaces/IBCoWPool.sol:117]: aside from that, nothing in the spec says we should externalize this value, so i guess we could remove the method alltogether

[the remark this answers is not in the document]
Comment (wei3erHase) [reply src/interfaces/IBCoWPool.sol:117]: https://hackmd.io/2_Xet-oyS6qjZWJZo7fD0A?view

File: src/interfaces/IBFactory.sol [hunk 10]
```diff
@@ -18,6 +18,16 @@ interface IBFactory {
...
 
+  /**
+   * @notice Thrown when caller is not BLabs address
+   */
+  error BFactory_NotBLabs();
+
+  /**
+   * @notice Thrown when the ERC20 transfer fails
+   */
+  error BFactory_ERC20TransferFailed();
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply src/interfaces/IBFactory.sol:29]: but BFactory is _not_ BPool

**Review by 0xAustrian — APPROVED** [review]

**Review by 0xteddybear — APPROVED** [review]

**Pull request merged**